### DISCUSSION TITLE
Update on the 'zopen' docs

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [zopencommunity/meta#414] [General]
I would like some input on improving the `zopen` docs. 

This is a re-write of what I had here before... 

Goals for docs:
- Don't repeat ourselves. Repetition means that we write it in 2 places and later on, only update one. Chaos ensues.
- The closer to the code the better. But no closer 😁 If the docs are in the code, then it is more likely to stay in sync with the code. No guarantees, but more likely. The tension with this item is that writing good docs in a programming language doesn't always work well.

With these 2 principles, I'd like to propose the following:
- Use --help to print out help, e.g. `zopen --help` prints out the help for the general zopen command. 
- Have sub-commands 'own' their own help, e.g. `zopen install --help` prints out the help for _zopen install_
  - As a side note, it would be good to have zopen-install supported as a first class script rather than being buried under 'lib'
- Utilize `help2man`, which already exists on z/OS, to generate man pages from the _--help_ output
- Find a tool to generate web pages (probably not markdown?) from man pages (presumably this exists - c.f. online man pages)

I believe this tackles most of the concerns/objections people brought up. 

@IgorTodorovskiIBM @DevonianTeuchter your thoughts appreciated.

**Comment by IgorTodorovskiIBM** [comment]
My preference would be to have a single source that describes the options. 
One other option would be to use help2man to generate the man page from the help output, and then a tool like pandoc to generate the markdown from the man page (or something like [help2md](https://github.com/valeriangalliat/help2md)).  If `man` doesn't exist, it can print out the help, but if it does, it can display the help2man output via man

**Comment by DevonianTeuchter** [comment]
I guess the fear is keeping the man pages up to date if it is a manual process but don't have strong feelings tbh though I'd be more inclined to veer the automation path...
Running help2man + pandoc within the build to autogenerate would seem to give us the best of all worlds in that the zopen scripts would be the source-of-truth for the help and would not need to determine if man was installed to display usable help [and without the user referring either to the internet or having man installed]. 
Using help2man would also mean the --help output would have to be well defined and standardised and doesn't appear to need many unnatural features :-)

**Comment by MikeFultonDev** [comment]
if we take the help2man approach, I have two concerns, which perhaps we can address:
- i want to be able to provide a fairly lengthy 'man' and 'web' solution but my hunch is when someone says 'git --help' they don't want 40 pages of stuff scrolling by on their screen. One idea that @IgorTodorovskiIBM and I discussed was that perhaps we could have an environment variable that, if specified, will print out a very detailed help when --help is specified. We could set this environment variable on when help2man is used, but not by default. 
- i want to be able to write in a fairly natural way when I'm writing the man pages, and be able to have the ability to have things like italics and so forth that I'd be able to get from a markdown page. I don't know if help2man gives me the ability to do this without the actual help looking horrible. Maybe it isn't a big deal because the 'extended' help that is written out when the environment variable is specified could look odd on the screen - but no one would actually do that! 

What do people think?

**Reply by IgorTodorovskiIBM** [reply]
I agree, --help should be more concise vs the man page. Another approach is if one specifies an extra option like --detailed-help, they get the full details.

I think the extra formatting would be an issue since I don't believe help2man supports that natively. 
However, we could provide conditionals for this as well. 

For example:
```
"--help provides help, visit $(genLink http://example.com)"
```
genLink would check if you're generating a help, man page or markdown and then generate the proper syntax.
```
<a href="http://example.com">http://example.com</a> # for markdown
```
Same for bold, italics, etc.

**Comment by MikeFultonDev** [comment]
I wonder if there is a 'help2md'? If we used something like that, then we could have the markdown in the extended help that could just 'pass through' to the md file (and then we could run pandoc to convert md to man pages? 

I expect the first 'syntax' part for help2man probably looks nice - so we probably don't need any markup in that?

**Reply by IgorTodorovskiIBM** [reply]
Yes, if it passes down the text verbatim, then that's a viable solution

**Reply by v1gnesh** [reply]
Putting a related thing out here - https://github.com/cpuguy83/go-md2man
There are other markdown-related libs too in the vicinity that may help flesh out what we want - https://github.com/bashup/mdsh

**Comment by MikeFultonDev** [comment]
no idea if this is any good: https://github.com/valeriangalliat/help2md/blob/master/README.md

**Comment by MikeFultonDev** [comment]
I tried out `help2man` on 'diff' to see how it works:
```
(rm -rf /tmp/man; mkdir -p /tmp/man/man1; help2man diff >/tmp/man/man1/diff.1; export MANPATH=/tmp/man; man diff)
```
 
creates a pretty nice man page:

![image](https://github.com/ZOSOpenTools/meta/assets/61100689/563528b9-b6ee-46c2-8f4f-148dfac5885d)
... (pruned stuff in the middle)
![image](https://github.com/ZOSOpenTools/meta/assets/61100689/0ce47acc-9901-4420-b1f1-98860c288017)

**Reply by MikeFultonDev** [reply]
So this seems to be 'pretty good' since there is also an option to include material from a side file, which we could use So we will want to provide help and version information for the different programs if we do this.
Perhaps we should create a forwarding script called 'zopen-xxx' for each of the verbs so that we can provide help (and so people can call them directly if they want?)

**Reply by MikeFultonDev** [reply]
If we go down this path, then I'd want to also have a 'help2md' which I'm thinking must exist - or alternately we do a 'man2md'. I'll poke around for this.

**Reply by IgorTodorovskiIBM** [reply]
Found this, it accepts overrides for the default help option:
```
EXECUTABLE should accept '--help' and '--version' options although alternatives may be specified using:

-h, --help-option=STRING
help option string
-v, --version-option=STRING
version option string
```
Also:
```
Additional material may be included in the generated output with the --include and --opt-include options. The format is simple:
```
This would be helpful for our case.

**Reply by MikeFultonDev** [reply]
removed my comments about --extended-help based on igor's comments

**Comment by DevonianTeuchter** [comment]
fwiw, the updates that went in for zopen-alt and zopen-clean lately should generate --help that can be consumed by help2man.
Having everything contained in the script itself keeps things organised so it could be that we have a --help-extended param that adds extra flourishes to the basic help text and is called by `help2man --help-option=help-extended`.

**Comment by MikeFultonDev** [comment]
NB: first experiments with help2man show that it can't tolerate escape characters (e.g. bold) so we'll have to remove that from our help
or restrict it to extended help.
I am ok removing the bold because it's not consistent with other --help outputs from tools 
@DevonianTeuchter any objections?

**Reply by DevonianTeuchter** [reply]
Not from me :-)

**Reply by MikeFultonDev** [reply]
@IgorTodorovskiIBM ?

**Reply by IgorTodorovskiIBM** [reply]
Good with me as well!

**Comment by MikeFultonDev** [comment]
Since `help2man` grabs interesting stuff from --version, I would like to propose we improve our --version a bit for zopen.
Here is `diff --version`

```
diff --version
diff (GNU diffutils) 3.8
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Paul Eggert, Mike Haertel, David Hayes,
Richard Stallman, and Len Tower.
```

I am thinking we could make the `zopen` set of tools be something like:

```
zopen [<verb>] --version
zopen-<verb> (zopen utils) V.R.M
Copyright (C) 2023 z/OS Open Tools Community
License Apache Version 2.0, January 2004 <https://www.apache.org/licenses>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by contributors to the z/OS Open Tools Community <https://github.com/ZOSOpenTools/meta/graphs/contributors>
```

Any concerns?

**Reply by IgorTodorovskiIBM** [reply]
I like it!

**Reply by MikeFultonDev** [reply]
zopen-install example:
```
zopen-install (zopen tools) 0.7.0
Copyright (C) 2023 z/OS Open Tools Community
License Apache Version 2.0, January 2004 <https://www.apache.org/licenses>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by contributors to the z/OS Open Tools Community <https://github.com/ZOSOpenTools/meta/graphs/contributors>
```

zopen example:
```
zopen (zopen tools) 0.7.0
Copyright (C) 2023 z/OS Open Tools Community
License Apache Version 2.0, January 2004 <https://www.apache.org/licenses>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by contributors to the z/OS Open Tools Community <https://github.com/ZOSOpenTools/meta/graphs/contributors>
```

**Reply by MikeFultonDev** [reply]
here is how zopen --version renders with help2man
![image](https://github.com/ZOSOpenTools/meta/assets/61100689/eeddfd52-0da8-49f2-ae8d-fe16b6c94471)

**Comment by MikeFultonDev** [comment]
Created an issue: https://github.com/ZOSOpenTools/meta/issues/534
